crates/wit-component/dl/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+//! This library uses lookup tables generated by `wasm-tools component link` to
+//! provide runtime library and symbol resolution via `dlopen` and `dlsym`.  The
+//! tables are provided via a call to `__wasm_set_libraries`, which `wasm-tools
+//! component link` arranges as part of component instantiation.
+
+use core::{
+    ffi::{c_char, c_int, c_void, CStr},
+    ptr, slice,
+};
+
+const RTLD_LAZY: c_int = 1;
+const RTLD_NOW: c_int = 2;
+
+const RTLD_NEXT: isize = -1;
+const RTLD_DEFAULT: isize = 0;
+
+#[repr(C)]
+pub struct Name {
+    length: u32,
+    data: *const u8,
+}
+
+#[repr(C)]
+pub struct Symbol {
+    name: Name,
+    address: *const c_void,
+}
+
+#[repr(C)]
+pub struct Symbols {
+    count: u32,
+    symbols: *const Symbol,
+}
+
+#[repr(C)]
+pub struct Library {
+    name: Name,
+    symbols: Symbols,
+}
+
+#[repr(C)]
+pub struct Libraries {
+    count: u32,
+    libraries: *const Library,
+}
+
+struct Pointer<T>(*const T);
+
+unsafe impl<T> Sync for Pointer<T> {}
+
+static mut ERROR: Pointer<c_char> = Pointer(ptr::null());
+static mut LIBRARIES: Pointer<Libraries> = Pointer(ptr::null());
+
+unsafe fn invalid_handle(library: *const c_void) -> bool {
+    if LIBRARIES.0.is_null() {
+        panic!(
+            "`__wasm_set_libraries` should have been called during \
+             instantiation with a non-NULL value"
+        );
+    }
+
+    let library = library as *const Library;
+    if (0..(*LIBRARIES.0).count).any(|index| {
+        (*LIBRARIES.0)
+            .libraries
+            .add(usize::try_from(index).unwrap())
+            == library
+    }) {
+        false
+    } else {
+        ERROR.0 = b"invalid library handle\0" as *const _ as _;
```

Comment:
While you're here you may be interested to try out `c"foo"` which I think will be stable in 1.78 but should already be stable on nightly